feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,364 @@
+package control_plane
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+
+	"sync"
+	"sync/atomic"
+	"time"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+
+	listener "github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	v2route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	xds "github.com/envoyproxy/go-control-plane/pkg/server"
+
+	"github.com/golang/glog"
+	"google.golang.org/grpc"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+
+	hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	orig_src "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/listener/original_src/v3"
+
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"magma/feg/cloud/go/protos"
+)
+
+type EnvoyController interface {
+	UpdateSnapshot([]*protos.AddUEHeaderEnrichmentRequest)
+}
+
+type ControllerClient struct {
+	version int32
+	config  cache.SnapshotCache
+}
+
+const (
+	XdsCluster                  = "xds_cluster"
+	Ads                         = "ads"
+	Xds                         = "xds"
+	Rest                        = "rest"
+	debug                       = true
+	port                        = 18000
+	gatewayPort                 = 18001
+	mode                        = Ads
+	any_addr                    = "0.0.0.0"
+	maxConcurrentStreams        = 16
+	initialStreamWindowSize     = 65536   // 64Kib
+	initialConnectionWindowSize = 1048576 // 1 MiB
+	connectTimeout              = 6 * time.Second
+	idleTimeout                 = 3600 * time.Second
+	httpPort                    = 80
+	clusterName                 = "cluster1"
+	targetPrefix                = "/"
+)
+
+type logger struct{}
+
+func (logger logger) Infof(format string, args ...interface{}) {
+	glog.Infof(format, args...)
+}
+func (logger logger) Errorf(format string, args ...interface{}) {
+	glog.Errorf(format, args...)
+}
+func (cb *callbacks) Report() {
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	glog.Infof("cb.Report() fetches %d,  callbacks %d", cb.fetches, cb.requests)
+}
+func (cb *callbacks) OnStreamOpen(ctx context.Context, id int64, typ string) error {
+	glog.Infof("OnStreamOpen %d open for %s", id, typ)
+	return nil
+}
+func (cb *callbacks) OnStreamClosed(id int64) {
+	glog.Infof("OnStreamClosed %d closed", id)
+}
+func (cb *callbacks) OnStreamRequest(int64, *v2.DiscoveryRequest) error {
+	glog.Infof("OnStreamRequest")
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	cb.requests++
+	if cb.signal != nil {
+		close(cb.signal)
+		cb.signal = nil
```

Comment:
looking at this it seems like we want to nil it to prevent panic? https://nanxiao.gitbooks.io/golang-101-hacks/content/posts/nil-channel-vs-closed-channel.html

Reply:
I am still not sure you need that. 

You can play with this
https://repl.it/@OriolBatalla/Closing-channels